### ISSUE TITLE
twisted.spread: Assume that Python always has unicode support enabled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7667]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7667|
|Type|enhancement|
|Created|2014-10-04 09:02:58Z|
|Branch|https://github.com/twisted/twisted/tree/jelly-require-unicode-7667|

this simplifies things for Python 3

Many other places in the assume this too, and
zope.interface also does (see interface/_compat.py)
    
Python 2.5 can still be compiled without unicode support,
Python 2.6 cannot
Python 2.7 can, see http://bugs.python.org/issue8767 (fix not applied to 2.6)

Attachments:

* [7667.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c17/c17e632ecba1d619ffaf2a43fd2804ca6352a097/c94782fc930fb47a0591159fc154ef7f7e1da77a.patch) (2519 bytes) - added by wolfgang61 on 2014-10-04 09:07:48Z - 
* [7667b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c17/c17e632ecba1d619ffaf2a43fd2804ca6352a097/e5c382c32b183adf72947281e49f243eccf3f490.patch) (2517 bytes) - added by wolfgang61 on 2014-10-04 22:07:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7667 7667
type__enhancement enhancement
reporter__wolfgang61 wolfgang61
priority__normal normal
milestone__ 
branch__branches_jelly_require_unicode_7667 branches/jelly-require-unicode-7667
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__spread spread
time__1412413378977371 1412413378977371
changetime__1412550235002637 1412550235002637
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to wolfgang61|
|-|-|

Thanks.

  1. Put the news fragment in `twisted/topfiles`
  1. The news fragment should be a `.removal` not a `.remove`
  1. Use fully qualified names in news fragments (and follow the style guidelines on ReviewProcess).
  1. For symmetry, it might be nice to `from twisted.python.compat import unicode as UnicodeType`.  This shortens the patch a little bit too.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wolfgang61 set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_1):
>   1. For symmetry, it might be nice to `from twisted.python.compat import unicode as UnicodeType`.  This shortens the patch a little bit too.

Yes it would be shorter but when porting to PY3 I also need to import unicode unrenamed, so both unicode and UnicodeType would exist, being identical and both imported from python.compat.unicode. I would prefer to avoid that.

When having to choose between unicode or UnicodeType everywhere, I believe unicode is easier to understand.

So this shorter patch would make the PY3 patch longer, and that is exactly what I want to avoid.

news file is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [43233]) Branching to 'jelly-require-unicode-7667'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [43234]) Apply 7667b.patch 

refs [#7667](https://github.com/twisted/twisted/issues/7667)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

It occurs to me that this is an incompatible change.  I hope no one is building Python without unicode support (that seems like something you might have wanted to do for a couple weeks in 2002) but who knows.  It would probably be best if you sent an email like https://twistedmatrix.com/pipermail/twisted-python/2014-September/028797.html to twisted-python.  The conditions for proceeding after that are described on <https://twistedmatrix.com/trac/wiki/CompatibilityPolicy#ProcedureforExceptionstothisPolicy>.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> It occurs to me that this is an incompatible change.

Actually, nevermind.  This supposes that Twisted actually works with unicode disabled.  But apart from the tiny steps the `twisted.spread.jelly` module takes to support this, there's no effort in Twisted to work without unicode.  And in fact not even `twisted.spread.jelly` works without unicode because its various dependencies don't.

So I'll just go ahead and merge this when buildbot is green.  Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|wolfgang61 commented|
|-|-|

Replying to [exarkun](#note_5):
> It occurs to me that this is an incompatible change.

So I was too fast, adding this comment anyway.

Not incompatible. zope.interface is unusable without Python unicode support. I already wrote that above but without proof. Here comes the proof.

After doing the tests below which are a quite strong indicator, I wanted to
be sure and compiled python 2.7.8 without unicode. This example shows that not even 
from zope import interface will work.

```
Python 2.7.8 (default, Oct  5 2014, 23:56:22) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> unicode
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'unicode' is not defined
>>> from zope import interface
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/zope/interface/__init__.py", line 57, in <module>
    _wire()
  File "/usr/local/lib/python2.7/site-packages/zope/interface/interface.py", line 696, in _wire
    from zope.interface.interfaces import IAttribute
  File "/usr/local/lib/python2.7/site-packages/zope/interface/interfaces.py", line 23, in <module>
    _BLANK = _u('')
  File "/usr/local/lib/python2.7/site-packages/zope/interface/_compat.py", line 22, in _u
    return unicode(s, 'unicode_escape')
NameError: global name 'unicode' is not defined
```

According to Twisted/setup.py, zope.interface >= 3.6.0 is required.

Zope.interface 3.6.0 uses unicode once without defining it:
```
def _normalize_name(name):
    if isinstance(name, basestring):
        return unicode(name)
```
unicode is not defined by zope.interface 3.6.0, so it expects it to be built in.

So I installed 3.6.0, placed an assert 0 there and did the following which IMHO is a very strong indicator that banana cannot be used without unicode support.


```
i5:~/ssdsrc/Twisted ((d42e3c1...)) python2.6
Python 2.6.8 (unknown, Oct  5 2014, 13:05:48) 
[GCC 4.8.2] on linux3
Type "help", "copyright", "credits" or "license" for more information.
>>> from twisted.spread.banana import LIST
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "twisted/spread/banana.py", line 17, in <module>
    from twisted.internet import protocol
  File "twisted/internet/protocol.py", line 568, in <module>
    interfaces.IConsumer)
  File "twisted/python/components.py", line 85, in registerAdapter
    factory = _registered(self, origInterface, interfaceClass)
  File "twisted/python/components.py", line 65, in _registered
    return registry.registered([required], provided)
  File "zope/interface/adapter.py", line 133, in registered
    name = _normalize_name(name)
  File "zope/interface/adapter.py", line 646, in _normalize_name
    assert 0
AssertionError
```

One more test, this time with zope-interface 4.0.5, again doing assert 0 before zope.interface uses unicode (which 4.0.5 never defines either):
```
Python 2.6.8 (unknown, Oct  5 2014, 13:05:48) 
[GCC 4.8.2] on linux3
Type "help", "copyright", "credits" or "license" for more information.
>>> from zope import interface
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "zope/interface/__init__.py", line 57, in <module>
    _wire()
  File "zope/interface/interface.py", line 696, in _wire
    from zope.interface.interfaces import IAttribute
  File "zope/interface/interfaces.py", line 23, in <module>
    _BLANK = _u('')
  File "zope/interface/_compat.py", line 22, in _u
    assert 0
AssertionError
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> Not incompatible. zope.interface is unusable without Python unicode support.

Thanks.  Sorry for making you revisit this.  What you wrote about zope.interface in the ticket description apparently didn't register.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [43237]) Merge jelly-require-unicode-7667

Author: wolfgang61
Reviewer: exarkun
Fixes: [#7667](https://github.com/twisted/twisted/issues/7667)

Remove the pointless Python-without-unicode support code from
`twisted.spread.jelly`.
